From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Christian=20K=C3=B6nig?= <christian.koenig@amd.com>
Date: Tue, 9 Nov 2021 11:08:18 +0100
Subject: dma-buf: specify usage while adding fences to dma_resv obj v7

Instead of distingting between shared and exclusive fences specify
the fence usage while adding fences.

Rework all drivers to use this interface instead and deprecate the old one.

v2: some kerneldoc comments suggested by Daniel
v3: fix a missing case in radeon
v4: rebase on nouveau changes, fix lockdep and temporary disable warning
v5: more documentation updates
v6: separate internal dma_resv changes from this patch, avoids to
    disable warning temporary, rebase on upstream changes
v7: fix missed case in lima driver, minimize changes to i915_gem_busy_ioctl

Signed-off-by: Christian König <christian.koenig@amd.com>
Reviewed-by: Daniel Vetter <daniel.vetter@ffwll.ch>
Link: https://patchwork.freedesktop.org/patch/msgid/20220407085946.744568-3-christian.koenig@amd.com

(Backported from commit 73511edf8b196e6f1ccda0fdf294ff57aa2dc9db linux)

Reviewed-by: Arvind Yadav <arvind.yadav@intel.com>
Signed-off-by: Varun Gupta <varun.gupta@intel.com>
---
 drivers/dma-buf/dma-resv.c | 85 ++++++++++++++++++++++++++++----------
 include/linux/dma-resv.h   | 32 ++++++++++----
 2 files changed, 86 insertions(+), 31 deletions(-)

diff --git a/drivers/dma-buf/dma-resv.c b/drivers/dma-buf/dma-resv.c
index 9ea42d37ced6..b78f8dcb6770 100644
--- a/drivers/dma-buf/dma-resv.c
+++ b/drivers/dma-buf/dma-resv.c
@@ -52,7 +52,7 @@
  */
 
 DEFINE_WD_CLASS(reservation_ww_class);
-EXPORT_SYMBOL(reservation_ww_class);
+EXPORT_SYMBOL_GPL(reservation_ww_class);
 
 /**
  * dma_resv_list_alloc - allocate fence list
@@ -106,7 +106,7 @@ void dma_resv_init(struct dma_resv *obj)
 	RCU_INIT_POINTER(obj->fence, NULL);
 	RCU_INIT_POINTER(obj->fence_excl, NULL);
 }
-EXPORT_SYMBOL(dma_resv_init);
+EXPORT_SYMBOL_GPL(dma_resv_init);
 
 /**
  * dma_resv_fini - destroys a reservation object
@@ -129,7 +129,7 @@ void dma_resv_fini(struct dma_resv *obj)
 	dma_resv_list_free(fobj);
 	ww_mutex_destroy(&obj->lock);
 }
-EXPORT_SYMBOL(dma_resv_fini);
+EXPORT_SYMBOL_GPL(dma_resv_fini);
 
 /**
  * dma_resv_reserve_shared - Reserve space to add shared fences to
@@ -206,18 +206,18 @@ int dma_resv_reserve_shared(struct dma_resv *obj, unsigned int num_fences)
 
 	return 0;
 }
-EXPORT_SYMBOL(dma_resv_reserve_shared);
+EXPORT_SYMBOL_GPL(dma_resv_reserve_shared);
 
 #ifdef CONFIG_DEBUG_MUTEXES
 /**
- * dma_resv_reset_shared_max - reset shared fences for debugging
+ * dma_resv_reset_max_fences - reset shared fences for debugging
  * @obj: the dma_resv object to reset
  *
  * Reset the number of pre-reserved shared slots to test that drivers do
  * correct slot allocation using dma_resv_reserve_shared(). See also
  * &dma_resv_list.shared_max.
  */
-void dma_resv_reset_shared_max(struct dma_resv *obj)
+void dma_resv_reset_max_fences(struct dma_resv *obj)
 {
 	struct dma_resv_list *fences = dma_resv_shared_list(obj);
 
@@ -227,7 +227,7 @@ void dma_resv_reset_shared_max(struct dma_resv *obj)
 	if (fences)
 		fences->shared_max = fences->shared_count;
 }
-EXPORT_SYMBOL(dma_resv_reset_shared_max);
+EXPORT_SYMBOL_GPL(dma_resv_reset_max_fences);
 #endif
 
 /**
@@ -238,7 +238,9 @@ EXPORT_SYMBOL(dma_resv_reset_shared_max);
  * Add a fence to a shared slot, obj->lock must be held, and
  * dma_resv_reserve_shared() has been called.
  */
-void dma_resv_add_shared_fence(struct dma_resv *obj, struct dma_fence *fence)
+static void __dma_resv_add_shared_fence(struct dma_resv *obj,
+					 struct dma_fence *fence,
+					 enum dma_resv_usage usage)
 {
 	struct dma_resv_list *fobj;
 	struct dma_fence *old;
@@ -254,10 +256,12 @@ void dma_resv_add_shared_fence(struct dma_resv *obj, struct dma_fence *fence)
 	write_seqcount_begin(&obj->seq);
 
 	for (i = 0; i < count; ++i) {
+		enum dma_resv_usage old_usage = DMA_RESV_USAGE_READ;
 
 		old = rcu_dereference_protected(fobj->shared[i],
-						dma_resv_held(obj));
-		if (old->context == fence->context ||
+					       dma_resv_held(obj));
+		if ((old->context == fence->context && old_usage >= usage &&
+		     (old == fence || dma_fence_is_later(fence, old))) ||
 		    dma_fence_is_signaled(old))
 			goto replace;
 	}
@@ -270,18 +274,24 @@ void dma_resv_add_shared_fence(struct dma_resv *obj, struct dma_fence *fence)
 	RCU_INIT_POINTER(fobj->shared[i], fence);
 	/* fence update must be visible before we extend the shared_count */
 	smp_wmb();
-	fobj->shared_count = count;
+	WRITE_ONCE(fobj->shared_count, count);
 
 	write_seqcount_end(&obj->seq);
 	dma_fence_put(old);
 }
-EXPORT_SYMBOL(dma_resv_add_shared_fence);
+
+void dma_resv_add_shared_fence(struct dma_resv *obj, struct dma_fence *fence)
+{
+	__dma_resv_add_shared_fence(obj, fence, DMA_RESV_USAGE_READ);
+}
+EXPORT_SYMBOL_GPL(dma_resv_add_shared_fence);
 
 /**
  * dma_resv_replace_fences - replace fences in the dma_resv obj
  * @obj: the reservation object
  * @context: the context of the fences to replace
  * @replacement: the new fence to use instead
+ * @usage: how the new fence is used, see enum dma_resv_usage
  *
  * Replace fences with a specified context with a new fence. Only valid if the
  * operation represented by the original fence has no longer access to the
@@ -291,12 +301,16 @@ EXPORT_SYMBOL(dma_resv_add_shared_fence);
  * update fence which makes the resource inaccessible.
  */
 void dma_resv_replace_fences(struct dma_resv *obj, uint64_t context,
-			     struct dma_fence *replacement)
+			     struct dma_fence *replacement,
+			     enum dma_resv_usage usage)
 {
 	struct dma_resv_list *list;
 	struct dma_fence *old;
 	unsigned int i;
 
+	/* Only readers supported for now */
+	WARN_ON(usage != DMA_RESV_USAGE_READ);
+
 	dma_resv_assert_held(obj);
 
 	write_seqcount_begin(&obj->seq);
@@ -320,7 +334,7 @@ void dma_resv_replace_fences(struct dma_resv *obj, uint64_t context,
 
 	write_seqcount_end(&obj->seq);
 }
-EXPORT_SYMBOL(dma_resv_replace_fences);
+EXPORT_SYMBOL_GPL(dma_resv_replace_fences);
 
 /**
  * dma_resv_add_excl_fence - Add an exclusive fence.
@@ -329,7 +343,8 @@ EXPORT_SYMBOL(dma_resv_replace_fences);
  *
  * Add a fence to the exclusive slot.  The obj->lock must be held.
  */
-void dma_resv_add_excl_fence(struct dma_resv *obj, struct dma_fence *fence)
+void dma_resv_add_excl_fence(struct dma_resv *obj,
+				    struct dma_fence *fence)
 {
 	struct dma_fence *old_fence = dma_resv_excl_fence(obj);
 	struct dma_resv_list *old;
@@ -358,7 +373,33 @@ void dma_resv_add_excl_fence(struct dma_resv *obj, struct dma_fence *fence)
 
 	dma_fence_put(old_fence);
 }
-EXPORT_SYMBOL(dma_resv_add_excl_fence);
+EXPORT_SYMBOL_GPL(dma_resv_add_excl_fence);
+
+void dma_resv_add_fence(struct dma_resv *obj, struct dma_fence *fence,
+		      enum dma_resv_usage usage)
+{
+	struct dma_resv_list *fobj;
+
+	dma_resv_assert_held(obj);
+
+	if (WARN_ON(!fence))
+		return;
+
+	fobj = dma_resv_shared_list(obj);
+	if (!fobj || fobj->shared_count >= fobj->shared_max) {
+		int err = dma_resv_reserve_shared(obj, 1);
+
+		if (err) {
+			pr_err("dma_resv: DROPPED fence %llu:%llu usage=%d on obj %p, reserve failed (%d)\n",
+			       fence->context, fence->seqno, usage, obj, err);
+			WARN_ON(1);
+			return;
+		}
+	}
+
+	__dma_resv_add_shared_fence(obj, fence, usage);
+}
+EXPORT_SYMBOL_GPL(dma_resv_add_fence);
 
 /**
  * dma_resv_iter_restart_unlocked - restart the unlocked iterator
@@ -437,7 +478,7 @@ struct dma_fence *dma_resv_iter_first_unlocked(struct dma_resv_iter *cursor)
 
 	return cursor->fence;
 }
-EXPORT_SYMBOL(dma_resv_iter_first_unlocked);
+EXPORT_SYMBOL_GPL(dma_resv_iter_first_unlocked);
 
 /**
  * dma_resv_iter_next_unlocked - next fence in an unlocked dma_resv obj.
@@ -462,7 +503,7 @@ struct dma_fence *dma_resv_iter_next_unlocked(struct dma_resv_iter *cursor)
 
 	return cursor->fence;
 }
-EXPORT_SYMBOL(dma_resv_iter_next_unlocked);
+EXPORT_SYMBOL_GPL(dma_resv_iter_next_unlocked);
 
 /**
  * dma_resv_iter_first - first fence from a locked dma_resv object
@@ -556,7 +597,7 @@ int dma_resv_copy_fences(struct dma_resv *dst, struct dma_resv *src)
 		}
 
 		dma_fence_get(f);
-		if (dma_resv_iter_is_exclusive(&cursor))
+		if (dma_resv_iter_usage(&cursor) == DMA_RESV_USAGE_WRITE)
 			excl = f;
 		else
 			RCU_INIT_POINTER(list->shared[list->shared_count++], f);
@@ -573,7 +614,7 @@ int dma_resv_copy_fences(struct dma_resv *dst, struct dma_resv *src)
 
 	return 0;
 }
-EXPORT_SYMBOL(dma_resv_copy_fences);
+EXPORT_SYMBOL_GPL(dma_resv_copy_fences);
 
 /**
  * dma_resv_get_fences - Get an object's shared and exclusive
@@ -737,8 +778,8 @@ EXPORT_SYMBOL_GPL(dma_resv_get_singleton);
  * Returns -ERESTARTSYS if interrupted, 0 if the wait timed out, or
  * greater than zer on success.
  */
-long dma_resv_wait_timeout(struct dma_resv *obj, unsigned int usage, bool intr,
-                           unsigned long timeout)
+long dma_resv_wait_timeout(struct dma_resv *obj, unsigned int usage,
+			   bool intr, unsigned long timeout)
 {
 	bool wait_all = !!usage;
 	long ret = timeout ? timeout : 1;
diff --git a/include/linux/dma-resv.h b/include/linux/dma-resv.h
index 5e68c0d83235..97adbfcab544 100644
--- a/include/linux/dma-resv.h
+++ b/include/linux/dma-resv.h
@@ -100,6 +100,9 @@ struct dma_resv_iter {
 	/** @fence: the currently handled fence */
 	struct dma_fence *fence;
 
+	/** @fence_usage: the usage of the current fence */
+	enum dma_resv_usage fence_usage;
+
 	/** @seq: sequence number to check for modifications */
 	unsigned int seq;
 
@@ -149,11 +152,17 @@ static inline void dma_resv_iter_end(struct dma_resv_iter *cursor)
 }
 
 /**
- * dma_resv_iter_is_exclusive - test if the current fence is the exclusive one
+ * dma_resv_iter_usage - Return the usage of the current fence
  * @cursor: the cursor of the current position
  *
- * Returns true if the currently returned fence is the exclusive one.
+ * Returns the usage of the currently processed fence.
  */
+static inline enum dma_resv_usage
+dma_resv_iter_usage(struct dma_resv_iter *cursor)
+{
+	return cursor->fence_usage;
+}
+
 static inline bool dma_resv_iter_is_exclusive(struct dma_resv_iter *cursor)
 {
 	return cursor->index == 0;
@@ -207,11 +216,13 @@ static inline bool dma_resv_iter_is_restarted(struct dma_resv_iter *cursor)
 #define dma_resv_assert_held(obj) lockdep_assert_held(&(obj)->lock.base)
 
 #ifdef CONFIG_DEBUG_MUTEXES
-void dma_resv_reset_shared_max(struct dma_resv *obj);
+void dma_resv_reset_max_fences(struct dma_resv *obj);
 #else
-static inline void dma_resv_reset_shared_max(struct dma_resv *obj) {}
+static inline void dma_resv_reset_max_fences(struct dma_resv *obj) {}
 #endif
 
+#define dma_resv_reset_shared_max dma_resv_reset_max_fences
+
 /**
  * dma_resv_lock - lock the reservation object
  * @obj: the reservation object
@@ -335,7 +346,7 @@ static inline struct ww_acquire_ctx *dma_resv_locking_ctx(struct dma_resv *obj)
  */
 static inline void dma_resv_unlock(struct dma_resv *obj)
 {
-	dma_resv_reset_shared_max(obj);
+	dma_resv_reset_max_fences(obj);
 	ww_mutex_unlock(&obj->lock);
 }
 
@@ -399,18 +410,21 @@ void dma_resv_init(struct dma_resv *obj);
 void dma_resv_fini(struct dma_resv *obj);
 int dma_resv_reserve_shared(struct dma_resv *obj, unsigned int num_fences);
 void dma_resv_add_shared_fence(struct dma_resv *obj, struct dma_fence *fence);
-void dma_resv_replace_fences(struct dma_resv *obj, uint64_t context,
-							struct dma_fence *fence);
 void dma_resv_add_excl_fence(struct dma_resv *obj, struct dma_fence *fence);
+void dma_resv_add_fence(struct dma_resv *obj, struct dma_fence *fence,
+			enum dma_resv_usage usage);
+void dma_resv_replace_fences(struct dma_resv *obj, uint64_t context,
+							struct dma_fence *fence,
+							enum dma_resv_usage usage);
 int dma_resv_get_fences(struct dma_resv *obj, struct dma_fence **pfence_excl,
 			unsigned *pshared_count, struct dma_fence ***pshared);
 int dma_resv_get_singleton(struct dma_resv *obj, enum dma_resv_usage usage,
 						  struct dma_fence **fence);
 int dma_resv_copy_fences(struct dma_resv *dst, struct dma_resv *src);
 
-long dma_resv_wait_timeout(struct dma_resv *obj, enum dma_resv_usage usage,
+long dma_resv_wait_timeout(struct dma_resv *obj, unsigned int usage,
 			   bool intr, unsigned long timeout);
-bool dma_resv_test_signaled(struct dma_resv *obj, enum dma_resv_usage usage);
+bool dma_resv_test_signaled(struct dma_resv *obj, unsigned int usage);
 
 static inline int dma_resv_reserve_fences(struct dma_resv *obj,
 					  unsigned int num_fences)
-- 
2.43.0

